### DISCUSSION TITLE
How to use an async Reader

### DISCUSSION THREAD

**Discussion by matthewpflueger** [marblejs/marble#372] [Q&A]
Hi,

I have a bunch of dependencies that I create using createReader. Something like:

```typescript
const dbClientDep = createReader(ask => {  
  return new dbClient(useContext(databaseURLToken)(ask))
})
```
Then later:
```typescript
bindEagerlyTo(dbClientToken)(dbClientDep)
```
This works great! But now I find I need to create a Reader whose handler returns a Promise, something like:
```typescript
const asyncDbClientDep = createReader(async (ask) => {
  return await asyncDBClientCreator(useContext(databaseURLToken)(ask))  
})
```
Then later:
```typescript
bindEagerlyTo(asyncDbClientToken)(asyncDbClientDep)
```
Except when later trying to use this dependency I always get back the Promise but what I expected was the resolved value.

I have other async dependencies like:

```typescript
bindEagerlyTo(anotherAsyncThing)(() => functionThatReturnsAPromise())
```
And it works well (I get the value of the resolved Promise not the Promise itself).

Is it possible to do something similar with createReader?

**Comment by JozefFlakus** [comment]
Interesting. It should work without any problems with `createReader`. I've prepared a small unit test #376  that double validates the scenario, including different forms of building async readers.

Could you write similar test case with a minimal dependency set in order to validate the behavior? I would like to narrow the registered context as much as possible, maybe there is something strange happening with your registered dependencies.
